src/k2/persistence/plog_client/PlogClient.cpp

```diff
@@ -0,0 +1,203 @@
+    }
+
+    return seastar::when_all_succeed(appendFutures.begin(), appendFutures.end())
+        .then([this, expected_offset, appended_offset](std::vector<std::tuple<Status, dto::PlogAppendResponse> >&& results) { 
+            Status return_status;
+            for (auto& result: results){
+                auto& [status, response] = result;
+                return_status = std::move(status);
+                if (!return_status.is2xxOK()) 
+                    break;
+                if (response.newOffset != expected_offset){
+                    return_status = Statuses::S500_Internal_Server_Error("offset inconsistent");
+                    break;
+                }
+            }
+            return seastar::make_ready_future<std::tuple<Status, uint32_t> >(std::tuple<Status, uint32_t>(std::move(return_status), std::move(expected_offset)));
+        });
+}
+
+
+seastar::future<std::tuple<Status, Payload>> PlogClient::read(String plogId, uint32_t offset, uint32_t size){
+    dto::PlogReadRequest request{.plogId = std::move(plogId), .offset=offset, .size=size};
+
+    return RPC().callRPC<dto::PlogReadRequest, dto::PlogReadResponse>(dto::Verbs::PERSISTENT_READ, request, *_persistenceMapEndpoints[_persistenceNameList[_persistenceMapPointer]][0], _plog_timeout()).
+        then([this] (auto&& result) {
+            auto& [status, response] = result;
+
+            return seastar::make_ready_future<std::tuple<Status, Payload> >(std::tuple<Status, Payload>(std::move(status), std::move(response.payload)));
+        });
+}
+
+seastar::future<std::tuple<Status, uint32_t>> PlogClient::seal(String plogId, uint32_t offset){
+    dto::PlogSealRequest request{.plogId = std::move(plogId), .truncateOffset=offset};
+
+    std::vector<seastar::future<std::tuple<Status, dto::PlogSealResponse> > > sealFutures;
+    for (auto& ep:_persistenceMapEndpoints[_persistenceNameList[_persistenceMapPointer]]){
+        sealFutures.push_back(RPC().callRPC<dto::PlogSealRequest, dto::PlogSealResponse>(dto::Verbs::PERSISTENT_SEAL, request, *ep, _plog_timeout()));
+    }
+
+    return seastar::when_all_succeed(sealFutures.begin(), sealFutures.end())
+        .then([this](std::vector<std::tuple<Status, dto::PlogSealResponse> >&& results) { 
+            Status return_status;
+            uint32_t sealed_offset;
+            for (auto& result: results){
+                auto& [status, response] = result;
+                return_status = std::move(status);
+                sealed_offset = response.sealedOffset;
+                if (!return_status.is2xxOK()) 
+                    break;
+            }
+            return seastar::make_ready_future<std::tuple<Status, uint32_t> >(std::tuple<Status, uint32_t>(std::move(return_status), std::move(sealed_offset)));
+        });
+}
+
+String PlogClient::_generatePlogId(){
+    String plogid = "TPCC_CLIENT_plog_0123456789";
+    std::mt19937 g(std::rand());

```

Comment:
I think it may be faster if you keep this as a member variable instead of creating a new one each time you want to generate an id

Reply:
we decided to fix this in a later pr